File: backend-springboot/src/main/java/com/splitwiseplus/controller/ExpenseController.java
package com.splitwiseplus.controller;

import com.splitwiseplus.model.Expense;
import com.splitwiseplus.model.User;
import com.splitwiseplus.model.Notification;
import com.splitwiseplus.model.Settlement;
import com.splitwiseplus.repository.ExpenseRepository;
import com.splitwiseplus.repository.UserRepository;
import com.splitwiseplus.repository.NotificationRepository;
import com.splitwiseplus.repository.SettlementRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.bson.types.ObjectId;
import java.util.List;

@RestController
@RequestMapping("/api/expenses")
@CrossOrigin(origins = "*") // Allow React to connect
public class ExpenseController {

    @Autowired
    private ExpenseRepository expenseRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private SettlementRepository settlementRepository;

    @PostMapping("/add")
    public Expense createExpense(@RequestBody Expense expense) {
        Expense savedExpense = expenseRepository.save(expense);

        // Notify split members
        if (savedExpense.getSplitWith() != null && !savedExpense.getSplitWith().isEmpty()) {
            User payer = null;
            if (savedExpense.getUserId() != null) {
                payer = userRepository.findById(savedExpense.getUserId().toString()).orElse(null);
            }
            String payerName = payer != null ? payer.getName() : savedExpense.getPaidBy();

            for (Expense.SplitMember item : savedExpense.getSplitWith()) {
                if (item.getEmail() == null || item.getEmail().isEmpty()) continue;

                userRepository.findByEmail(item.getEmail()).ifPresent(targetUser -> {
                    // Don't notify the payer themselves
                    if (savedExpense.getUserId() == null || !targetUser.getId().equals(savedExpense.getUserId().toString())) {
                        Notification notification = new Notification();
                        notification.setRecipientId(new ObjectId(targetUser.getId()));
                        notification.setSenderId(savedExpense.getUserId());
                        notification.setType("expense_added");
                        notification.setMessage(String.format("%s added a new expense: \"%s\". You owe ₹%.2f.", 
                            payerName, savedExpense.getTitle(), item.getAmount()));
                        notification.setExpenseId(new ObjectId(savedExpense.getId()));
                        notificationRepository.save(notification);
                    }
                });
            }
        }

        return savedExpense;
    }

    @GetMapping("/{userId}")
    public List<Expense> getExpensesByUser(@PathVariable String userId) {
        ObjectId userObjectId = new ObjectId(userId);
        User user = userRepository.findById(userId).orElse(null);
        String email = user != null ? user.getEmail() : "___no_email___";
        String name = user != null ? user.getName() : "___no_name___";
        
        return expenseRepository.findByUserIdOrEmailOrName(userObjectId, email, name);
    }

    @DeleteMapping("/{id}")
    public void deleteExpense(@PathVariable String id) {
        expenseRepository.deleteById(id);
    }

    @PutMapping("/{id}")
    public Expense updateExpense(@PathVariable String id, @RequestBody Expense expense) {
        expense.setId(id);
        return expenseRepository.save(expense);
    }

    @GetMapping("/settle/{userId}")
    public List<Settlement> getSettlements(@PathVariable String userId) {
        ObjectId userObjectId = new ObjectId(userId);
        User user = userRepository.findById(userId).orElse(null);
        String name = user != null ? user.getName() : "___no_name___";
        
        return settlementRepository.findByUserIdOrPersonName(userObjectId, name);
    }

    @PostMapping("/settle")
    public Settlement createSettlement(@RequestBody Settlement settlement) {
        Settlement savedSettlement = settlementRepository.save(settlement);

        // Notify the original payer that someone settled
        if (savedSettlement.getExpenseId() != null) {
            expenseRepository.findById(savedSettlement.getExpenseId().toString()).ifPresent(expense -> {
                User settler = null;
                if (savedSettlement.getUserId() != null) {
                    settler = userRepository.findById(savedSettlement.getUserId().toString()).orElse(null);
                }
                String settlerName = settler != null ? settler.getName() : savedSettlement.getPersonName();

                Notification notification = new Notification();
                notification.setRecipientId(expense.getUserId());
                notification.setSenderId(savedSettlement.getUserId());
                notification.setType("expense_settled");
                notification.setMessage(String.format("%s marked an expense as settled: \"%s\" (₹%.2f).", 
                    settlerName, expense.getTitle(), savedSettlement.getAmount()));
                notification.setExpenseId(expense.getUserId()); // This should be expense.getId() converted to ObjectId
                // Wait, in Node it was expense._id
                notification.setExpenseId(new ObjectId(expense.getId()));
                
                notificationRepository.save(notification);
            });
        }

        return savedSettlement;
    }

    @DeleteMapping("/settle/delete/{id}")
    public void deleteSettlement(@PathVariable String id) {
        settlementRepository.deleteById(id);
    }
}

File: backend/controllers/expenseController.js
const Expense = require("../models/Expense")
const Settlement = require("../models/Settlement")
const User = require("../models/User")
const Notification = require("../models/Notification")

// ADD EXPENSE
exports.addExpense = async (req, res) => {
  try {
    const expense = await Expense.create(req.body)

    // Notify split members
    if (expense.splitWith && expense.splitWith.length > 0) {
      const payer = await User.findById(expense.userId)
      const payerName = payer ? payer.name : expense.paidBy

      for (const item of expense.splitWith) {
        if (!item.email) continue

        const targetUser = await User.findOne({ email: item.email })
        if (targetUser && targetUser._id.toString() !== expense.userId.toString()) {
          await Notification.create({
            recipientId: targetUser._id,
            senderId: expense.userId,
            type: "expense_added",
            message: `${payerName} added a new expense: "${expense.title}". You owe ₹${item.amount.toLocaleString()}.`,
            expenseId: expense._id
          })
        }
      }
    }

    res.status(201).json(expense)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// GET EXPENSES BY USER
exports.getExpenses = async (req, res) => {
  try {
    const user = await User.findById(req.params.userId);
    const userEmail = user ? user.email : "___no_email___";
    const userName = user ? user.name : "___no_name___";

    const expenses = await Expense.find({
      $or: [
        { userId: req.params.userId },
        { "splitWith.email": userEmail },
        { "splitWith.name": userName } // Fallback for legacy data
      ]
    })
    res.json(expenses)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// DELETE
exports.deleteExpense = async (req, res) => {
  try {
    await Expense.findByIdAndDelete(req.params.id)
    res.json({ message: "Deleted" })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// UPDATE
exports.updateExpense = async (req, res) => {
  try {
    const updated = await Expense.findByIdAndUpdate(
      req.params.id,
      req.body,
      { new: true }
    )
    res.json(updated)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// SETTLE PAYMENT
exports.settlePayment = async (req, res) => {
  try {
    const { expenseId, personName, amount, userId } = req.body
    const settlement = await Settlement.create({
      expenseId,
      personName,
      amount,
      userId
    })

    // Notify the original payer that someone settled
    const expense = await Expense.findById(expenseId)
    if (expense) {
      const settler = await User.findById(userId)
      const settlerName = settler ? settler.name : personName

      await Notification.create({
        recipientId: expense.userId,
        senderId: userId,
        type: "expense_settled",
        message: `${settlerName} marked an expense as settled: "${expense.title}" (₹${amount.toLocaleString()}).`,
        expenseId: expense._id
      })
    }

    res.status(201).json(settlement)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// GET SETTLEMENTS BY USER
exports.getSettlements = async (req, res) => {
  try {
    const user = await User.findById(req.params.userId);
    const settlements = await Settlement.find({
      $or: [
        { userId: req.params.userId },
        { personName: user ? { $regex: new RegExp(`^${user.name}$`, "i") } : "___" }
      ]
    })
    res.json(settlements)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// DELETE SETTLEMENT
exports.deleteSettlement = async (req, res) => {
  try {
    await Settlement.findByIdAndDelete(req.params.id)
    res.json({ message: "Settlement deleted" })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}